scrapypuppeteer/middleware/restore.py

```diff
@@ -0,0 +1,157 @@
+    RESTORING_LENGTH_SETTING = "RESTORING_LENGTH"
+
+    def __init__(self, restoring_length: int, n_retry_restoring: int):
+        self.restoring_length = restoring_length
+        self.n_retry_restoring = n_retry_restoring
+        self.context_requests = {}
+        self.context_length = {}
+
+    @classmethod
+    def from_crawler(cls, crawler: Crawler):
+        restoring_length = crawler.settings.get(cls.RESTORING_LENGTH_SETTING, 1)
+        if not isinstance(restoring_length, int):
+            raise TypeError(
+                f"`{cls.RESTORING_LENGTH_SETTING}` must be an integer, got {type(restoring_length)}"
+            )
+        elif restoring_length < 1:
+            raise ValueError(
+                f"`{cls.RESTORING_LENGTH_SETTING}` must be greater than or equal to 1"
+            )
+
+        n_retry_restoring = crawler.settings.get(cls.N_RETRY_RESTORING_SETTING, 1)
+        if not isinstance(n_retry_restoring, int):
+            raise TypeError(
+                f"`{cls.N_RETRY_RESTORING_SETTING}` must be an integer, got {type(n_retry_restoring)}"
+            )
+        elif n_retry_restoring < 1:
+            raise ValueError(
+                f"`{cls.N_RETRY_RESTORING_SETTING}` must be greater than or equal to 1"
+            )
+
+        return cls(restoring_length, n_retry_restoring)
+
+    @staticmethod
+    def process_request(request, spider):
+        if not isinstance(request, PuppeteerRequest):
+            return None
+
+        if not request.meta.pop("recover_context", False):
+            return None
+
+        if request.context_id or request.page_id:
+            raise IgnoreRequest(
+                f"Request {request} is not in the beginning of the request-response sequence"
+            )
+
+        request.meta["__request_binding"] = True
+        request.dont_filter = True
+        return None
+
+    def process_response(self, request, response, spider):
+        puppeteer_request: Union[PuppeteerRequest, None] = request.meta.get(
+            "puppeteer_request", None
+        )
+        request_binding = puppeteer_request is not None and puppeteer_request.meta.get(
+            "__request_binding", False
+        )
+
+        if isinstance(response, PuppeteerResponse):
+            if request_binding:
+                self._bind_context(request, response)
+            if response.context_id in self.context_length:
+                # Update number of actions in context
+                self.context_length[response.context_id] += 1
+        elif (
+            puppeteer_request is not None
+            and response.status == HTTPStatus.UNPROCESSABLE_ENTITY
+        ):
+            # One PuppeteerRequest has failed with 422 error
+            if request_binding:
+                # Could not get context, retry
+                if request.meta.get("__restore_count", 0) < self.n_retry_restoring:
+                    request.meta["__restore_count"] += 1
+                    return request
```

Comment:
I'm not sure if this situation even possible.
I could not reproduce it, but I added it here.
What do you think?

Reply:
Still not sure